File: client/runtime/wasm/abi.ts
// GoSX direct runtime ABI validation. Constants come exclusively from the
// generated runtime contract included immediately before this source.
// @ts-check

/**
 * @typedef {object} GoSXRuntimeHandshakeProbe
 * @property {number} abiVersion
 * @property {number} featureMask
 * @property {string} variant
 * @property {number} mailboxVersion
 * @property {string} manifestHash
 */

(function() {
  "use strict";
  if (typeof window === "undefined") return;

  const contract = window.__gosx_runtime_contract;
  if (!contract) {
    console.error("[gosx] generated runtime ABI contract is missing");
    return;
  }

  /** @param {string} variant */
  function featureMaskForVariant(variant) {
    const name = String(variant || "");
    return Number(contract.variants[name] || (contract.compatibilityVariants && contract.compatibilityVariants[name]) || 0) >>> 0;
  }

  /**
   * @param {number} required
   * @param {object} [published]
   */
  function selectVariant(required, published) {
    const mask = Number(required || 0) >>> 0;
    const candidates = Object.keys(contract.variants).filter(function(variant) {
      const features = featureMaskForVariant(variant);
      return (features & mask) === mask && (!published || published[variant]);
    });
    if (published) {
      candidates.sort(function(left, right) {
        const leftSize = Number(published[left] && published[left].size) || Number.MAX_SAFE_INTEGER;
        const rightSize = Number(published[right] && published[right].size) || Number.MAX_SAFE_INTEGER;
        if (leftSize !== rightSize) return leftSize - rightSize;
        return left < right ? -1 : left > right ? 1 : 0;
      });
    }
    return candidates[0] || "";
  }

  function referenceError(runtimeRef) {
    if (!runtimeRef || typeof runtimeRef.path !== "string" || runtimeRef.path === "") return "runtime path is missing";
    if (!/^[0-9a-f]{16,64}$/i.test(String(runtimeRef.hash || ""))) return "runtime asset hash is missing or invalid";
    const expectedFeatures = featureMaskForVariant(runtimeRef.variant);
    if (!expectedFeatures) return "runtime variant is missing or unknown";
    if ((Number(runtimeRef.featureMask) >>> 0) !== expectedFeatures) return "runtime reference feature mask does not match its variant";
    if (String(runtimeRef.manifestHash || "") !== contract.manifestHash) return "runtime manifest identity is missing or stale";
    return "";
  }

  /**
   * @param {GoSXRuntimeHandshakeProbe|null|undefined} handshake
   * @param {object|number} [expected]
   */
  function validationError(handshake, expected) {
    if (!handshake) return "runtime handshake is missing";
    if (Number(handshake.abiVersion) !== contract.abiVersion) return "runtime ABI version mismatch";
    if (Number(handshake.mailboxVersion) !== contract.mailboxVersion) return "runtime mailbox version mismatch";
    const declaredFeatures = featureMaskForVariant(handshake.variant);
    if (!declaredFeatures) return "runtime handshake variant is unknown";
    if ((Number(handshake.featureMask) >>> 0) !== declaredFeatures) return "runtime handshake feature mask does not match its variant";
    if (String(handshake.manifestHash || "") !== contract.manifestHash) return "runtime handshake manifest identity mismatch";

    if (expected && typeof expected === "object") {
      const refError = referenceError(expected);
      if (refError) return refError;
      if (String(handshake.variant) !== String(expected.variant)) return "runtime handshake variant does not match the manifest reference";
      if ((Number(handshake.featureMask) >>> 0) !== (Number(expected.featureMask) >>> 0)) return "runtime handshake feature mask does not match the manifest reference";
      if (String(handshake.manifestHash) !== String(expected.manifestHash)) return "runtime handshake manifest identity does not match the manifest reference";
      return "";
    }

    const required = Number(expected || 0) >>> 0;
    if ((declaredFeatures & required) !== required) return "runtime is missing required capabilities";
    return "";
  }

  gosxRuntime.support = {
    abiVersion: contract.abiVersion,
    mailboxVersion: contract.mailboxVersion,
    manifestHash: contract.manifestHash,
    featureMaskForVariant: featureMaskForVariant,
    selectVariant: selectVariant,
    referenceError: referenceError,
    validationError: validationError,
    validateHandshake: function(handshake, expected) { return validationError(handshake, expected) === ""; },
    requiredFeatureMask: function(required) { return Number(required || 0) >>> 0; },
  };
})();

File: client/runtime/wasm/abi.go
// Package wasm contains the versioned browser/WASM wire contract.
//
// The package deliberately has no syscall/js dependency.  The same constants
// and codecs are used by the host-side tests, the TinyGo entry point, and the
// generated TypeScript contract, so a browser can reject an incompatible
// runtime before it starts sending hot-path messages.
package wasm

import (
	"fmt"
	"strings"
)

const (
	// ABIVersion changes whenever the direct browser/WASM contract changes in a
	// way that cannot be handled by a compatibility reader.
	ABIVersion uint32 = 2

	// MailboxVersion is independent from the export ABI.  It lets the mailbox
	// codec evolve while the direct function surface remains compatible.
	MailboxVersion uint16 = 1
)

// FeatureMask identifies the capabilities compiled into one runtime variant.
type FeatureMask uint32

const (
	FeatureCore FeatureMask = 1 << iota
	FeatureEngine
	FeatureCollab
	FeatureScene3D
	FeatureIslands
)

// Variant is the stable name carried by the handshake and manifest.
type Variant string

const (
	VariantCore    Variant = "core"
	VariantEngine  Variant = "engine"
	VariantCollab  Variant = "collab"
	VariantFull    Variant = "full"
	VariantIslands Variant = "islands"
)

// Handshake is returned by the direct runtime export before the host selects a
// mailbox path.  ManifestHash is optional for development builds and required
// by canonical production evidence.
type Handshake struct {
	ABIVersion     uint32      `json:"abiVersion"`
	FeatureMask    FeatureMask `json:"featureMask"`
	Variant        Variant     `json:"variant"`
	MailboxVersion uint16      `json:"mailboxVersion"`
	ManifestHash   string      `json:"manifestHash,omitempty"`
}

// NewHandshake creates the runtime declaration for a compiled variant.
func NewHandshake(variant Variant, manifestHash string) Handshake {
	return Handshake{
		ABIVersion:     ABIVersion,
		FeatureMask:    FeatureMaskForVariant(variant),
		Variant:        variant,
		MailboxVersion: MailboxVersion,
		ManifestHash:   strings.TrimSpace(manifestHash),
	}
}

// FeatureMaskForVariant returns the closed capability set for a variant.
// Later variants may add bits, but they must never repurpose an existing bit.
func FeatureMaskForVariant(variant Variant) FeatureMask {
	switch variant {
	case VariantCore:
		return FeatureCore
	case VariantEngine:
		return FeatureCore | FeatureEngine | FeatureScene3D
	case VariantCollab:
		return FeatureCore | FeatureCollab | FeatureIslands
	case VariantIslands:
		return FeatureCore | FeatureIslands
	case VariantFull:
		return FeatureCore | FeatureEngine | FeatureCollab | FeatureScene3D | FeatureIslands
	default:
		return 0
	}
}

// Validate checks the compatibility boundary before any runtime operation.
func (h Handshake) Validate(required FeatureMask) error {
	if h.ABIVersion != ABIVersion {
		return fmt.Errorf("runtime ABI mismatch: got %d, want %d", h.ABIVersion, ABIVersion)
	}
	if h.MailboxVersion != MailboxVersion {
		return fmt.Errorf("runtime mailbox mismatch: got %d, want %d", h.MailboxVersion, MailboxVersion)
	}
	if h.Variant == "" || FeatureMaskForVariant(h.Variant) == 0 {
		return fmt.Errorf("runtime handshake has unknown variant %q", h.Variant)
	}
	if h.FeatureMask&required != required {
		return fmt.Errorf("runtime %q is missing feature mask 0x%x", h.Variant, uint32(required&^h.FeatureMask))
	}
	return nil
}

// Supports reports whether the handshake exposes every requested feature.
func (h Handshake) Supports(required FeatureMask) bool {
	return h.Validate(required) == nil
}
